File: apps/api/src/domain/entities/school.py
from dataclasses import dataclass, field
from datetime import UTC, datetime
from uuid import UUID, uuid4


class Role:
    STUDENT = "student"
    TEACHER = "teacher"
    SCHOOL_ADMIN = "school_admin"
    PARENT = "parent"

    STAFF = frozenset({TEACHER, SCHOOL_ADMIN})
    ALL = frozenset({STUDENT, TEACHER, SCHOOL_ADMIN, PARENT})


@dataclass
class School:
    name: str
    id: UUID = field(default_factory=uuid4)
    slug: str = ""
    address: str | None = None
    created_by: UUID | None = None
    is_active: bool = True
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))


@dataclass
class SchoolMember:
    school_id: UUID
    user_id: UUID
    id: UUID = field(default_factory=uuid4)
    role: str = Role.TEACHER
    status: str = "active"
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    @property
    def is_school_admin(self) -> bool:
        return self.role == Role.SCHOOL_ADMIN


@dataclass
class Classroom:
    school_id: UUID
    name: str
    id: UUID = field(default_factory=uuid4)
    subject: str | None = None
    grade: str | None = None
    teacher_id: UUID | None = None
    join_code: str = ""
    description: str | None = None
    is_active: bool = True
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))


@dataclass
class Enrollment:
    classroom_id: UUID
    student_id: UUID
    id: UUID = field(default_factory=uuid4)
    status: str = "active"
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    @property
    def is_active(self) -> bool:
        return self.status == "active"


@dataclass
class SyllabusItem:
    classroom_id: UUID
    subject: str
    chapter: str
    id: UUID = field(default_factory=uuid4)
    topic: str | None = None
    description: str | None = None
    order_index: int = 0
    knowledge_base_id: UUID | None = None
    is_published: bool = False
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))

File: apps/native/src/api/school.ts
import { apiClient } from './client'

export interface Classroom {
  id: string
  school_id: string
  name: string
  subject: string | null
  grade: string | null
  teacher_id: string | null
  join_code: string
  description: string | null
  is_active: boolean
  student_count: number
  syllabus_count: number
  created_at: string
  updated_at: string
}

export interface School {
  id: string
  name: string
  slug: string
  address: string | null
  is_active: boolean
  my_role: string
  created_at: string
  updated_at: string
}

export interface SyllabusItem {
  id: string
  classroom_id: string
  subject: string
  chapter: string
  topic: string | null
  description: string | null
  order_index: number
  knowledge_base_id: string | null
  is_published: boolean
  created_at: string
  updated_at: string
}

export interface RosterEntry {
  enrollment_id: string
  student_id: string
  username: string
  email: string
  status: string
  enrolled_at: string
}

export const schoolApi = {
  // Student
  joinClassroom: (joinCode: string) =>
    apiClient.post<Classroom>('/api/v1/school/classrooms/join', { join_code: joinCode }),

  getEnrolled: () =>
    apiClient.get<Classroom[]>('/api/v1/school/classrooms/enrolled/mine'),

  getSyllabus: (classroomId: string) =>
    apiClient.get<SyllabusItem[]>(`/api/v1/school/classrooms/${classroomId}/syllabus`),

  // Teacher
  getMySchools: () => apiClient.get<School[]>('/api/v1/school/schools/mine'),

  createSchool: (name: string) =>
    apiClient.post<School>('/api/v1/school/schools', { name }),

  getMyClassrooms: () =>
    apiClient.get<Classroom[]>('/api/v1/school/classrooms/mine'),

  createClassroom: (payload: {
    school_id: string
    name: string
    subject?: string | null
    grade?: string | null
  }) => apiClient.post<Classroom>('/api/v1/school/classrooms', payload),

  getRoster: (classroomId: string) =>
    apiClient.get<RosterEntry[]>(`/api/v1/school/classrooms/${classroomId}/roster`),

  addSyllabusItem: (
    classroomId: string,
    payload: { subject: string; chapter: string; topic?: string | null; is_published?: boolean },
  ) => apiClient.post<SyllabusItem>(`/api/v1/school/classrooms/${classroomId}/syllabus`, payload),

  toggleSyllabusPublish: (itemId: string, isPublished: boolean) =>
    apiClient.patch<SyllabusItem>(`/api/v1/school/syllabus/${itemId}`, {
      is_published: isPublished,
    }),
}
